CUDA-Q Docs Applications Section Reorg

mawolf2023 opened the pull request:
<!--
Thanks for helping us improve CUDA-Q!

⚠️ The pull request title should be concise and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

Checklist:
- [x ] I have added tests to cover my changes.
- [ x] I have updated the documentation accordingly.
- [ x] I have read the CONTRIBUTING document.
-->

### Description
This PR updates the applications section of the docs. Notebooks can now be tagged and dynamically filtered.  Rather than a  generic icon, each application has a description and a preview image.

Made minor fixes to backends simulator table

Updated simulator diagram to include noisy simulators

Fixed double link to backends page.

Added CUDA-QX Page

khalatepradnya commented:
/ok to test

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,31 +1,205 @@
+        </div>
+        
+        <div class="tag-filters" style="margin-top: 20px;">
+            <h3>Specialized Backends:</h3>
```
I would add a 'Simulation' backend tag andhave the following as backend filters: Simulation, QPU, Noisy. If we can have a sub tag, then whenever someone clicks on simulation they can further select single / multi GPU / Multi QPU and when they select QPU they can select a sub tag of modality and/or specific provider. 
If the experience of the sub tags is too complex, we can add the single / multi GPU / Multi QPU and specific QPU/modality tags to the applications but limit backend filter options to only Simulation, QPU, Noisy.

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,31 +1,205 @@
+            <button class="tag-button" data-group="domain" data-tag="chemistry">Chemistry</button>
+            <button class="tag-button" data-group="domain" data-tag="fundamental">Fundamental Algorithms</button>
+            <button class="tag-button" data-group="domain" data-tag="aiforq">AI Supercomputing for Quantum</button>
+            <button class="tag-button" data-group="domain" data-tag="qforai">Quantum for AI</button>
```
Is there anything else here than QML? If not then maybe rename the tag to QML?

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,31 +1,205 @@
+        </div>
+
+        <div class="tag-filters" style="margin-top: 20px;">
+            <h3>CUDA-X Library:</h3>
```
I'm not sure if we need that, in the same way that we won't add a sample/observe tags, it's an implementation detail, the assumption is that the applications are written in the best way to utilize libraries and CUDA-Q APIs. I would add a QEC tag to the domain filters

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,31 +1,205 @@
+        </div>
+
+        <div class="tag-filters" style="margin-top: 20px;">
+            <h3>Filter by Collaboration:</h3>
```
Maybe rename to 'community'

khalatepradnya commented:
/ok to test

mawolf2023 commented:
/ok to test

mawolf2023 commented:
/ok to test

khalatepradnya commented:
/ok to test

mawolf2023 commented:
/ok to test

khalatepradnya commented on docs/sphinx/applications/python/H 0-pyscf.log:
```diff
@@ -0,0 +1,229 @@
+#INFO: **** input file is /home/cudaq/.local/lib/python3.10/site-packages/cudaq_solvers/bin/cudaq-pyscf ****
```
Can you confirm that this file needs to be checked into source tree?

khalatepradnya replied:
Same comment for `docs/sphinx/applications/python/H 0-pyscf.chk`

mawolf2023 replied:
Hmm, I am not sure why that is there.  It is not even clear to me which notebook that is sued in.  I removed all solvers implementations so it seems like that is in a solvers file?

khalatepradnya replied:
I've removed them.

khalatepradnya commented on docs/sphinx/applications/python/O 0-pyscf.log:
```diff
@@ -0,0 +1,238 @@
+#INFO: **** input file is /home/cudaq/.local/lib/python3.10/site-packages/cudaq_solvers/bin/cudaq-pyscf ****
```
And these files - `docs/sphinx/applications/python/O 0-pyscf.chk`, `docs/sphinx/applications/python/O 0-pyscf.log`

khalatepradnya commented on docs/sphinx/applications/python/qaoa.ipynb:
```diff
@@ -26,7 +26,41 @@
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
```
This seems to be an error message. 
Is it better to check in notebooks with output cleared?

mawolf2023 replied:
Hmm, I do not get an error when I run the notebook. Maybe that is better if errors keep showing up, but I also think seeing the output is helpful.

khalatepradnya replied:
Reran and fixed.

khalatepradnya commented on docs/sphinx/applications/python/qaoa.ipynb:
```diff
@@ -336,20 +370,27 @@
      "output_type": "stream",
      "text": [
-      "CUDA-Q Version latest (https://github.com/NVIDIA/cuda-quantum a726804916fd397408cbf595ce6fe5f33dcd8b4c)\n"
+      "CUDA-Q Version 0.9.1 (https://github.com/NVIDIA/cuda-quantum a3ab722eab7ba39ec81424278789fabca726c436)\n"
```
Should this change be checked into the source tree. I expect when the notebook is run and this cell is executed, the version would be updated to reflect the appropriate version.

mawolf2023 replied:
maybe, or maybe we just comment that out.  I am not sure the purpose of that.

khalatepradnya commented on docs/sphinx/conf.py:
```diff
@@ -48,10 +48,12 @@
     #'sphinx.ext.linkcode',
     'sphinx_reredirects',
     'breathe',
+    'sphinx_tags',
```
Note: This requires `pip install sphinx-tags` before running the docs gen script. It needs to be part of the dev deps image or at least be installed as part of the script.

khalatepradnya replied:
I've added to devdeps docker script.

khalatepradnya commented on docs/sphinx/using/backends/simulators.rst:
```diff
@@ -26,48 +26,54 @@ technical details and code examples for using each circuit simulator.
      - General purpose (default); Trajectory simulation for noisy circuits
      - Single GPU
-     - single / double
+     - single / double (default)
```
Is this a typo? Single precision is default as per rest of our docs.

mawolf2023 replied:
@efratshabtai Can you confirm?

efratshabtai replied:
@1tnguyen can you please go over the entire table and confirm?

khalatepradnya replied:
For @1tnguyen:
<img width="730" alt="image" src="https://github.com/user-attachments/assets/cb737a67-78dd-4ac5-a7fa-11ca417f8517" />

1tnguyen replied:
For the `nvidia` simulator, the default is `single` precision. This applies to `nvidia`, `nvidia, option=mgpu`, and `nvidia, option=mqpu`
Other than that, this table looks great 👍

khalatepradnya commented on .github/workflows/config/spelling_allowlist.txt:
```diff
@@ -99,6 +99,7 @@ QPU
 QPUs
 QPU’s
 QTX
+QX
```
For `CUDA-QX`.

khalatepradnya commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,233 @@
-    /applications/python/adapt_qaoa.ipynb
+====================
+
+:spellcheck-disable:
```
Spell checker is still complaining about the notebook names...

khalatepradnya replied:
Renamed notebooks to make spell checker happy.

khalatepradnya commented:
/ok to test

khalatepradnya commented:
/ok to test

khalatepradnya commented on docs/sphinx/_static/custom.css:
```diff
@@ -0,0 +1,157 @@
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *                                                                             *
+ * File: custom.css                                                            *
```
We don't have similar files in the source tree, so I don't know if the copyright banner format is same or different.
@schweitzpgi - can you provide pointers on the getting this one right for `*.css` and `.*.js` files?

schweitzpgi replied:
I think the banners are prefabbed. So you'll probably want this extra stuff in its own comment.

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+        <div class="tag-filters" style="margin-top: 20px;">
+    <h3>Filter by Backend:</h3>
+    <div class="backend-group">
+        <button class="tag-button backend-toggle" data-tag="noiseless">Noiseless Simulator</button>
```
Should we add CPU? (do we have any apps that run on CPU only?)

mawolf2023 replied:
@efratshabtai The only ones that might fit here are the super basic BV, DJ, QFT.  I would not expect this list to grow too much.  But we can add it if you think its absence would turn away some users. 

Alternatively, we could just add a note that any with the `nvidia` backend will automatically run on CPU if a GPU is unavailable, but the performance will vary by application.

efratshabtai replied:
@mawolf2023 where do you suggest adding this note?
My concern with the original comment was that a user that has cpu only will think they won't be able to run any of these apps.

mawolf2023 replied:
@efratshabtai I would be OK adding a tag for the few that are "CPU" if that is a concern.  

-BV
-QFT
-Quantum Teleportation
-DJ

There are probably more that could run on CPU, but I would just label these as the ones we are confident with.

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    <div class="notebook-entry" data-tags="optimization,noiseless,gpu">
+        <a href="/applications/python/QAOA.html" class="notebook-title">QAOA for Max Cut Problem</a>
+        <div class="notebook-content">
+            This notebook explains the theory behind the Quantum Approximate Optimization Algorithm (QAOA) and explains how it can be used to solve the Max Cut problem.
```
```suggestion
            This notebook explains the theory behind the Quantum Approximate Optimization Algorithm (QAOA) and how it can be used to solve the Max Cut problem.
```

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    <div class="notebook-entry" data-tags="chemistry,optimization,noiseless,gpu">
+        <a href="/applications/python/digitized_counterdiabatic_QAOA.html" class="notebook-title">Digitized Counterdiabatic QAOA</a>
+        <div class="notebook-content">
+            This notebook explores application of DC-QAOA to predict molecules that might be good candidates for drugs based on their interactions with proteins.
```
I like the descriptions starting in 'Learn...' of the other notebooks, maybe we can use it here and in the previous one too.

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    <div class="notebook-entry" data-tags="chemistry,qec,qpu,neutral,collab">
+        <a href="/applications/python/logical_aim_Sqale.html" class="notebook-title">The Anderson Impurity Model With Logical Qubits</a>
+        <div class="notebook-content">
+            A recent collaboration between NVIDIA and Infleqtion demonstrated a logical qubit workflow built in CUDA-Q and executed on the Infleqtion's neutral atom QPU.
```
```suggestion
            A collaboration between NVIDIA and Infleqtion demonstrated a logical qubit workflow built in CUDA-Q and executed on the Infleqtion's neutral atom QPU.
```

efratshabtai replied:
Can link to the paper: https://arxiv.org/abs/2412.07670

khalatepradnya replied:
Keeping 'unresolved' as reminder to add link to the paper.

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    <div class="notebook-entry" data-tags="fundamental,noiseless,gpu">
+        <a href="/applications/python/quantum_teleportation.html" class="notebook-title">Quantum Teleportation</a>
+        <div class="notebook-content">
+            Quantum teleportation is one of the strange phenomena that makes quantum computing so interesting. Learn how teleportation works and how it can be implemented in CUDA-Q.
```
```suggestion
            Quantum teleportation is one of the strange phenomena that makes quantum computing so interesting. Learn how teleportation works and how it is implemented in CUDA-Q.
```

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    <div class="notebook-entry" data-tags="fundamental,noisy,density">
+        <a href="/applications/python/quantum_volume.html" class="notebook-title">Quantum Volume</a>
+        <div class="notebook-content">
+            Benchmarking the performance of quantum computers, especially between different qubit modalities, is challenging. One method is to experimentally perform the quantum volume test. Learn how this test is performed and how it can be implemented in CUDA-Q.
```
```suggestion
            Benchmarking the performance of quantum computers, especially between different qubit modalities, is challenging. One method is to experimentally perform the quantum volume test. Learn how this test is performed and how it is implemented in CUDA-Q.
```

efratshabtai commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+        <img src="/_static/app_title_images/quantum_transformer_preview.png" alt="Transformer Preview" class="notebook-image">
+    </div>
+
+    <div class="notebook-entry" data-tags="chemistry,noiseless,gpu,collab">
```
We plan to remove this notebook in another PR as it is failing with latest code, will replace with updated code in a couple of weeks.

khalatepradnya replied:
Removed by merging with `main`.

khalatepradnya commented:
/ok to test

mawolf2023 commented:
@sacpis @efratshabtai  Can we still slide in an update here?  If not I can open another PR, but I realized one of the QPU logos was missing in the backends section. 
![qpus (2)](https://github.com/user-attachments/assets/7d9b3c2a-dbcf-40e8-b782-7017219fca04)

mawolf2023 commented on docs/sphinx/using/applications.rst:
```diff
@@ -1,33 +1,231 @@
+    </div>
+
+.. raw:: html
+
```
@sacpis Hmmm, Probably the best spot would be a linespace at the end of the description and then list the tags.

mawolf2023 commented:
@sacpis  I added the CPU tags for some applications.  I also added entries in applications.rst for quantum edge detection and ADAPT VQE.  I do not think these notebooks are currently in the CUDA-Q docs, but I just made sure they have a title image and description for when they merge.  Thanks!

1tnguyen reviewed (approved):
👍
